### PR TITLE
Allow a triplet to set `VCPKG_TARGET_` values

### PR DESCRIPTION
**Pull request by donny-dont**
In `vcpkg_common_definitions.cmake` the values of `VCPKG_TARGET_` are always set. Since the file is included after the triplet file there's no way for the triplet to specify those values without them being overwritten.

Add a macro `vcpkg_set_if_not_defined` which checks if `out_var` is already `DEFINED` and if not sets the it to `default`. Use this macro in `vcpkg_common_definitions` whenever a default value is set to prevent a triplet's value from being quashed.

### CODE REVIEW

**Comment by Neumann-A** [comment]
Not the right approach. What is your actual problem?

**Comment by donny-dont** [comment]
Hey @Neumann-A so the issue I encountered was when I was using `vcpkg_copy_tools` in a private portfile which uses a custom triplet for an embedded device. It wasn't finding the tool because it has a suffix, `.elf` and by default `VCPKG_TARGET_EXECUTABLE_SUFFIX` is the empty string. My first thought was oh okay I'll just set the value in the triplet file but then noticed that because of the order of includes, triplet first then `vcpkg_copy_tools` that the value I set was always being overridden. So my thought was to use `VCPKG_TARGET_EXECUTABLE_SUFFIX` as a sentinel value and if it was set then its assumed that the whole block of values is set in the triplet file.

**Comment by Neumann-A** [comment]
A better approach is probably do define a macro/function called `vcpkg_set_if_not_defined`. Which sets the value if the user did not provide an override for it.

**Comment by donny-dont** [comment]
Created the `macro(vcpkg_set_if_not_defined)` and used it whenever a default was being set in `vcpkg_common_definitions.cmake`. Let me know if that's what you had in mind @Neumann-A

**Comment by donny-dont** [comment]
The build breaks related to the change @Neumann-A ?

File: scripts/cmake/vcpkg_common_definitions.cmake [hunk 1]
```diff
@@ -104,13 +104,13 @@ elseif(VCPKG_TARGET_IS_OSX)
...
-    set(VCPKG_TARGET_STATIC_LIBRARY_PREFIX "lib")
-    set(VCPKG_TARGET_SHARED_LIBRARY_PREFIX "lib")
-    set(VCPKG_FIND_LIBRARY_SUFFIXES ".so" ".a")
-    set(VCPKG_FIND_LIBRARY_PREFIXES "lib" "")
+    vcpkg_set_if_not_defined(VCPKG_TARGET_STATIC_LIBRARY_SUFFIX ".a")
+    vcpkg_set_if_not_defined(VCPKG_TARGET_IMPORT_LIBRARY_SUFFIX "")
+    vcpkg_set_if_not_defined(VCPKG_TARGET_SHARED_LIBRARY_SUFFIX ".so")
+    vcpkg_set_if_not_defined(VCPKG_TARGET_STATIC_LIBRARY_PREFIX "lib")
+    vcpkg_set_if_not_defined(VCPKG_TARGET_SHARED_LIBRARY_PREFIX "lib")
+    vcpkg_set_if_not_defined(VCPKG_FIND_LIBRARY_SUFFIXES ".so" ".a")
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_common_definitions.cmake:112]: this has two args

File: scripts/cmake/vcpkg_set_if_not_defined.cmake [hunk 2]
```diff
@@ -0,0 +1,5 @@
+macro(vcpkg_set_if_not_defined out_var default)
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_set_if_not_defined.cmake:1]: this only as one. You need to use ARGN/ARGV correctly.

**Comment by jimwang118** [comment]
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

**Comment by jimwang118** [comment]
@donny-dont  Thanks for your PR, Is work still being done for this PR?

**Comment by jimwang118** [comment]
Closing this PR since it seems that no progress is being made. Please ping us to reopen if work is still being done.

**Pull request closed without merging**